spark_log_parser/parsing_models/application_model.py

```diff
@@ -241,29 +217,39 @@ def __init__(self, eventlogpath, bucket=None, stdoutpath=None, debug=False):  #
                         )
 
                 elif event_type == "SparkListenerExecutorAdded":
-                    hosts.append(json_data["Executor Info"]["Host"])
-                    self.cores_per_executor = int(json_data["Executor Info"]["Total Cores"])
-                    self.num_executors = self.num_executors + 1
+                    executor_id = json_data["Executor ID"]
+                    executor_info = json_data["Executor Info"]
+
+                    executor = self.executors[executor_id]
+                    executor.id = executor_id
+                    executor.start_time = json_data["Timestamp"]
+                    executor.host = executor_info["Host"]
+                    executor.cores = int(executor_info["Total Cores"])
+
+                    hosts.add(executor.host)
+
+                    # TODO - what happens if we receive multiple of these events and they have different
+                    #  amounts of cores?
```

Comment:
It seems a reasonable assumption most of the time to assume that Executors will have the same # of cores, but this of course may not always be true...

Reply:
The only time where I've see this isn't true is on EMR when a certain configuration is set: [spark.yarn.heterogeneousExecutors.enabled=true](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-5320-release.html#:~:text=Starting%20with%20Amazon%20EMR%205.32.0%20and%206.5.0%2C%20dynamic%20executor%20sizing%20for%20Apache%20Spark%20is%20enabled%20by%20default.%20To%20turn%20this%20feature%20on%20or%20off%2C%20you%20can%20use%20the%20spark.yarn.heterogeneousExecutors.enabled%20configuration%20parameter.). We saw this once in a log and decided to just reject this config because the predictor assumes homogeneous executors.